Introducing: Projects (#8930)
PrefectHQ/prefect: docs/tutorials/projects.md

In this example, we extend the above two by dockerizing our setup and executing runs with a Docker Worker.  Building off the [git-based example above](#git-based-deployment), let's switch our deployment to submit work to the `docker-work` work pool that we started at the beginning:

<div class="terminal">
```bash
$ prefect deploy -f 'log-flow' \
    -n my-docker-git-deployment \
    -p docker-work
$ prefect deployment run 'log-flow/my-docker-git-deployment'
```
</div>

As promised above, this worked out of the box!  

Let's deploy a new flow from this project that requires additional dependencies that might not be available in the default image our work pool is using; this flow requires both `pandas` and `numpy` as a dependency, which we will also install locally first and confirm the flow is working:

<div class="terminal">
```bash
$ pip install -r requirements.txt
$ python flows/pandas_flow.py
```
</div>

We now have two options for how to manage these dependencies in our worker's environment:

- setting the `EXTRA_PIP_PACKAGES` environment variable or using another hook to install the dependencies at runtime
- building a custom Docker image with the dependencies baked in

> The first option is the quickest, but it does run the risk of changing runtime state from one run to the next and is not recommended for production use cases.  For the second option, we need to configure a `build` step within our `prefect.yaml` file as follows (Note: if starting from scratch we could use the `docker-git` recipe):

Reviewer: It felt like whiplash to be told about EXTRA_PIP_PACKAGES and then to be told it's not good.  Can we switch to just focus on what we're doing in the tutorial.  E.g.:

```suggestion
In this tutorial we will focus on building a custom Docker image. First, we need to configure a `build` step within our `prefect.yaml` file as follows (Note: if starting from scratch we could use the `docker-git` recipe):
```

Author: love that suggestion, yup

- in this case, we are using `dockerfile: auto` to tell Prefect to automatically create a `Dockerfile` for us; otherwise we could write our own and pass its location as a path to the `dockerfile` kwarg
- to avoid dealing with real image registries, we are not pushing this image; in most use cases you will want `push: true` (which is the default)

All that's left to do is create our deployment and specify our image name to instruct the worker what image to pull:

<div class="terminal">
```bash
$ prefect deploy -f 'pandas-flow' \
    -n docker-build-deployment \
    -p docker-work \
    -v image=local-only/testing:dev
$ prefect deployment run 'log-flow/my-docker-git-deployment'
```
</div>

Your run should complete succsessfully, logs and all!  Note that the `-v` flag represents a job variable, which are the allowed pieces of infrastructure configuration on a given work pool.  Each work pool can customize the fields they accept here.

!!! tip "Templating values"
    As a matter of best practice, you should avoid hardcoding the image name and tag in both your `prefect.yaml` and CLI. Instead, you should [use variable templating](/concepts/projects/#templating-options).

> #### A counterexample

Reviewer 2: Totally agree, though the title confused me. At first I thought it was retroactively arguing that the previous section wasn't correct for some reason. I think it could be more precise ("Converting local deployments to Docker") or something

> #### A counterexample

Author: @jlowin you clearly don't read enough math

> #### A counterexample

Reviewer 3: I would call it "Failing early and explicitly with workers" which I think its the primary value prop you are explaining.

Revisiting [our local deployment above](#local-deployment), let's begin by switching it to submit work to our `docker-work` work pool by re-running `prefect deploy` to see what happens:

<div class="terminal">
```bash
$ prefect deploy ./api_flow.py:call_api \
    -n my-second-deployment \
    -p docker-work
$ prefect deployment run 'Call API/my-second-deployment'
```
</div>

This fails with the following error:
<div class="terminal">
```bash
ERROR: FileNotFoundError: [Errno 2] No such file or directory: '/Users/chris/dev/my-first-project'
```
</div>

The reason this occurs is because our deployment has a fundamentally local `pull` step; inspecting `prefect.yaml` we find:

```yaml
pull:
- prefect.projects.steps.set_working_directory:
    directory: /Users/chris/dev/my-first-project
```

In order to successfully submit such a project to a dockerized environment, we need to either:

> - [`push` this project](/concepts/projects/#the-push-section) to a remote location (such as GitHub, or possibly a Cloud storage bucket)

Author: Good call, not sure why I jumped to that haha

```bash
$ prefect project init
```
</div>

Note that you can safely run this command in a non-empty directory where you already have work, as well.

This command will create your `.prefect/` directory along with the two YAML files `deployment.yaml` and `prefect.yaml`; if any of these files or directories already exist, they will not be altered or overwritten.

!!! tip "Project Recipes"
    Prefect ships with multiple project recipes, which allow you to initialize a project with a more opinionated structure.  You can see all available recipes by running:

    <div class="terminal">
    ```bash
    $ prefect project recipe ls
    ```
    </div>

    And you can use recipes with the `--recipe` flag:

    <div class="terminal">
    ```bash
    $ prefect project init --recipe docker
    ```
    </div>

    If no recipe is provided, the `init` command makes an intelligent choice of recipe based on local configuration; for example, if you initialize a project within a git repository, Prefect will automatically use the `git` recipe.

## Creating a basic deployment

Projects are most useful for creating deployments; let's walk through some examples right now.  

### Local deployment

> In this example, we'll create a project from scratch that runs locally.  Let's start by creating a new directory and making that our working directory and initializing a project:

Reviewer 3: ```suggestion
In this example, we'll create a project from scratch that runs locally.  Let's start by creating a new directory, making that our working directory, and initializing a project:
```

    <div class="terminal">
    ```bash
    $ prefect project register-flow ./api_flow.py:call_api
    ```
    </div>

    This pre-registration allows you to deploy based on name instead of entrypoint path:
    <div class="terminal">
    ```bash
    $ prefect deploy -f 'Call API' \
        -n my-first-deployment \
        -p local-work
    ```
    </div>

### Git-based deployment

In this example, we'll initialize a project from [a pre-built GitHub repository](https://github.com/PrefectHQ/hello-projects) and see how it is automatically portable across machines.

We start by cloning the remote repository and initializing a project within the root of the repo directory:

<div class="terminal">
```bash
$ git clone https://github.com/PrefectHQ/hello-projects
$ cd hello-projects
$ prefect project init --recipe git
```
</div>

We can now proceed with the same steps as above to create a new deployment:

<div class="terminal">
```bash
$ prefect deploy -f 'log-flow' \
    -n my-git-deployment \
    -p local-work
```
</div>

> Notice that we were able to deploy based on flow name alone; this is because the repository owner [pre-registered the `log-flow` for us](https://github.com/PrefectHQ/hello-projects/blob/main/.prefect/flows.json).  Alternatively, we could run `prefect deploy ./flows/log_flow.py:log_flow` if we knew the full entrypoint path.

Reviewer 3: ```suggestion
Notice that we were able to deploy based on flow name alone; this is because the repository owner [pre-registered the `log-flow` for us](https://github.com/PrefectHQ/hello-projects/blob/main/.prefect/flows.json).  Alternatively, if we knew the full entrypoint path, we could run `prefect deploy ./flows/log_flow.py:log_flow`.
```

- Lastly, note the top line "Cloning into 'hello-projects'..."; because this project is based out of a GitHub repository, it is _automatically_ portable to any remote location where both `git` and `prefect` are configured! You can convince yourself of this by either running a new local worker on a different machine, or by switching this deployment to run with your docker work pool (more on this shortly).

!!! note "`prefect.yaml`"
    The above process worked out-of-the-box because of the information stored within `prefect.yaml`; if you open this file up in a text editor, you'll find that is not empty.  Specifically, it contains the following `pull` step that was automatically populated when you first ran `prefect project init`:
    ```yaml
    pull:
    - prefect.projects.steps.git_clone_project:
        repository: https://github.com/PrefectHQ/hello-projects.git
        branch: main
    ```
    These `pull` steps are the instructions sent to your worker's runtime environment that allow it to clone your project in remote locations. For more information, see [the project concept documentation](/concepts/projects/).

### Dockerized deployment

> In this example, we extend the above two by dockerizing our setup and executing runs with a Docker Worker.  Building off the [git-based example above](#git-based-deployment), let's switch our deployment to submit work to the `docker-work` work pool that [we started at the beginning](#worker-tip):

Reviewer 3: ```suggestion
In this example, we extend the examples above by dockerizing our setup and executing runs with a Docker Worker.  Building off the [git-based example above](#git-based-deployment), let's switch our deployment to submit work to the `docker-work` work pool that [we started at the beginning](#worker-tip):
```

<div class="terminal">
```bash
$ prefect deploy -f 'log-flow' \
    -n my-docker-git-deployment \
    -p docker-work
$ prefect deployment run 'log-flow/my-docker-git-deployment'
```
</div>

As promised above, this worked out of the box!  

> Let's deploy a new flow from this project that requires additional dependencies that might not be available in the default image our work pool is using; this flow requires both `pandas` and `numpy` as a dependency, which we will also install locally first and confirm the flow is working:

Reviewer 3: ```suggestion
Let's deploy a new flow from this project that requires additional dependencies that might not be available in the default image our work pool is using; this flow requires both `pandas` and `numpy` as a dependency, which we will install locally first to confirm the flow is working:
```

<div class="terminal">
```bash
$ pip install -r requirements.txt
$ python flows/pandas_flow.py
```
</div>

We now have two options for how to manage these dependencies in our worker's environment:

- setting the `EXTRA_PIP_PACKAGES` environment variable or using another hook to install the dependencies at runtime
- building a custom Docker image with the dependencies baked in

> In this tutorial we will focus on building a custom Docker image. First, we need to configure a `build` step within our `prefect.yaml` file as follows (Note: if starting from scratch we could use the `docker-git` recipe):

Reviewer 3: ```suggestion
In this tutorial, we will focus on building a custom Docker image. First, we need to configure a `build` step within our `prefect.yaml` file as follows (Note: if starting from scratch we could use the `docker-git` recipe):
```
